llrt_modules/src/modules/fs/file_handle.rs

```diff
@@ -0,0 +1,559 @@
+        }
+        Ok(())
+    }
+
+    async fn chown(&self, ctx: Ctx<'_>, uid: u32, gid: u32) -> Result<()> {
+        #[cfg(unix)]
+        {
+            let path = self.path.clone();
+            task::spawn_blocking(move || std::os::unix::fs::chown(&path, Some(uid), Some(gid)))
+                .await
+                .or_throw(&ctx)?
+                .or_throw_msg(&ctx, "Can't modify file owner")?;
+        }
+        Ok(())
+    }
+
+    async fn close(&mut self) {
+        if let Some(file) = self.file.take() {
+            drop(file.into_std().await);
+        }
+    }
+
+    async fn datasync(&self, ctx: Ctx<'_>) -> Result<()> {
+        self.file(&ctx)?
+            .sync_data()
+            .await
+            .or_throw_msg(&ctx, "Can't sync file data")?;
+        Ok(())
+    }
+
+    #[qjs(get)]
+    async fn fd(&self, ctx: Ctx<'_>) -> Result<isize> {
+        #[cfg(unix)]
+        {
+            use std::os::fd::AsRawFd;
+            self.file(&ctx)?.as_raw_fd().try_into().or_throw(&ctx)
+        }
+        #[cfg(windows)]
+        {
+            use std::os::windows::io::AsRawHandle;
+            let handle = self.file(&ctx)?.as_raw_handle();
+            Ok(handle as isize)
+        }
+        #[cfg(not(any(unix, windows)))]
+        {
+            Ok(0)
+        }
+    }
+
+    async fn read<'js>(
+        &mut self,
+        ctx: Ctx<'js>,
+        buffer_or_options: Opt<Either<ArrayBufferView<'js>, ReadOptions<'js>>>,
+        options_or_offset: Opt<Either<ReadOptions<'js>, usize>>,
+        length: Opt<usize>,
+        // Not supporting position for now since it is not available in tokio.
+        // See https://github.com/tokio-rs/tokio/issues/699
+        // position: Opt<Either<isize, Null>>,
+    ) -> Result<Object<'js>> {
+        let options_1 = match buffer_or_options.0 {
+            Some(Either::Left(buffer)) => ReadOptions {
+                buffer: Some(buffer),
+                ..Default::default()
+            },
+            Some(Either::Right(options)) => options,
+            None => ReadOptions::default(),
+        };
+        let options_2 = match options_or_offset.0 {
+            Some(Either::Left(options)) => options,
+            Some(Either::Right(offset)) => ReadOptions {
+                offset: Some(offset),
+                ..Default::default()
+            },
+            None => ReadOptions::default(),
+        };
+
+        let buffer = options_1
+            .buffer
+            .or(options_2.buffer)
+            .unwrap_or_else_ok(|| {
+                ArrayBufferView::from_buffer(&ctx, Buffer::alloc(DEFAULT_BUFFER_SIZE))
+            })?;
+        let offset = options_1.offset.or(options_2.offset).unwrap_or(0);
+        let length = options_1
+            .length
+            .or(options_2.length)
+            .or(length.0)
+            .unwrap_or_else(|| buffer.len() - offset);
+
+        // It is not safe to pass the buffer from `ArrayBufferView` to `File::read`
```

Comment:
This will not do because the buffer would need to be a fixed size so if we read more data than is required by the caller it would be lost if the next read is not executed on the same thread.

Considering we use a `mut self`, the user cannot call read twice on the same filehandle currently so we essentially just need a custom version of BufReader that never bypasses the internal buffer.

Reply:
>This will not do because the buffer would need to be a fixed size so if we read more data than is required by the caller it would be lost if the next read is not executed on the same thread.
You're right, this isn't feasible.
> Considering we use a mut self, the user cannot call read twice on the same filehandle currently so we essentially just need a custom version of BufReader that never bypasses the internal buffer.
Yeah we might need that

Reply:
I did some tests without an intermediate buffer and it seemed fine. You either get all of A file or all of B file. 
I think that in this particular case, it is """ok""" to pass the buffer directly since the API doesn't say anything about UB.

Reply:
@richarddavison Did you want me to keep the intermediate buffer or write directly to the pointer from quickjs?

Reply:
Let’s keep it as is for now and we can improve later. Generally we try to do the most performant operation even if it leads to a bit complex non-idiomatic code